library/sha256.c

```diff
@@ -37,6 +37,33 @@
 #if defined(__aarch64__)
 #  if defined(MBEDTLS_SHA256_USE_A64_CRYPTO_IF_PRESENT) || \
     defined(MBEDTLS_SHA256_USE_A64_CRYPTO_ONLY)
+/* *INDENT-OFF* */
+#    if !defined(__ARM_FEATURE_CRYPTO)
+#      if defined(__clang__)
+#        if __clang_major__ < 18
+           /* TODO: Re-consider above after https://reviews.llvm.org/D131064
+            *       merged.
+            *
+            * The intrinsic declaration are guarded with ACLE predefined macros
+            * in clang, and those macros are only enabled with command line.
+            * Define the macros can enable those declaration and avoid compile
+            * error on it.
+            */
+#          define __ARM_FEATURE_CRYPTO 1
+#        endif
+#        pragma clang attribute push (__attribute__((target("crypto"))), apply_to=function)
+#        define MBEDTLS_POP_TARGET_PRAGMA
+#      elif defined(__GNUC__)
+#        if __GNUC__ < 6 /* TODO: check sha256 compatible for GCC */
+#          error "A more recent GCC is required for MBEDTLS_SHA256_USE_A64_CRYPTO_*"
+#        else
+#          pragma GCC target ("arch=armv8-a+crypto")
```

Comment:
[follow-up] This is useful in the following situation: `MBEDTLS_SHA256_USE_A64_CRYPTO_IF_PRESENT` is enabled, and the user wants to build a library that works both on systems with the crypto extensions and systems without the crypto extensions. Then the code guarded by `mbedtls_a64_crypto_sha256_has_support()` should be compiled for the `armv8-a+crypto` architecture, and the rest of the code (including the pure-software SHA256 implementation) should be compiled for a more generic target.

Since AFAIK there's no way to undo a `#pragma GCC target` (no way to know the original target), that means the +crypto-specific code should be in its own file, separate from the software implementation.

It's currently impossible to build the library with a hybrid target like this, so this pull request is a step in the right direction, but it's incomplete. We can do the next step in a separate PR.

Reply:
Another point though: does it really make sense to change the build target under `MBEDTLS_SHA256_USE_A64_CRYPTO_ONLY`? If you compile for generic aarch64 and run the library on a machine without the crypto extensions, SHA-256 calculations will fault. Ok, technically, this could be something somebody wants: a program that uses SHA-256 only if it's accelerated, and SHA-512 otherwise (for performance), and that doesn't want to build the SHA-256 software implementation to save code size. I can see this making sense in some kind of integrity monitor. But this would be a pretty unusual scenario, so I think it's more likely to be a mistake.

So, open question: should we allow or forbid this target switching? The user can work around it by specifying different build flags when building `sha256.o`, for those unusual use cases where it's desired.

Reply:
> does it really make sense to change the build target under MBEDTLS_SHA256_USE_A64_CRYPTO_ONLY?

That makes sense for getting best optimization. But this case MUST compile library and application together with same flags.